classes/manager.php

```diff
@@ -0,0 +1,570 @@
+
+        foreach ($DB->get_records('tiny_elements_variant', ['categoryname' => $categoryname]) as $variant) {
+            $this->delete_variant($variant->id);
+        }
+
+        // Purge CSS and JS cache.
+        \tiny_elements\local\utils::purge_css_cache();
+        \tiny_elements\local\utils::rebuild_css_cache();
+        \tiny_elements\local\utils::purge_js_cache();
+        \tiny_elements\local\utils::rebuild_js_cache();
+    }
+
+    /**
+     * Delete a flavor.
+     *
+     * @param int $id
+     */
+    public function delete_flavor(int $id): void {
+        global $DB;
+        $sql = 'DELETE FROM {tiny_elements_comp_flavor}
+                WHERE flavorname IN (
+                    SELECT name FROM {tiny_elements_flavor}
+                    WHERE id = ?
+                )';
+        $DB->execute($sql, [$id]);
+        $DB->delete_records('tiny_elements_flavor', ['id' => $id]);
+
+        // Purge CSS cache.
+        \tiny_elements\local\utils::purge_css_cache();
+        \tiny_elements\local\utils::rebuild_css_cache();
+    }
+
+    /**
+     * Delete a variant.
+     *
+     * @param int $id
+     * @return void
+     */
+    public function delete_variant(int $id): void {
+        global $DB;
+        $sql = 'DELETE FROM {tiny_elements_comp_variant}
+                WHERE variant IN (
+                    SELECT name FROM {tiny_elements_variant}
+                    WHERE id = ?
+                )';
+        $DB->execute($sql, [$id]);
+        $DB->delete_records('tiny_elements_variant', ['id' => $id]);
+
+        // Purge CSS cache.
+        \tiny_elements\local\utils::purge_css_cache();
+        \tiny_elements\local\utils::rebuild_css_cache();
+    }
+
+    /**
+     * Delete a component.
+     *
+     * @param int $id
+     * @return void
+     */
+    public function delete_component(int $id): void {
+        global $DB;
+        $componentname = $DB->get_field('tiny_elements_component', 'name', ['id' => $id]);
+        $DB->delete_records('tiny_elements_comp_flavor', ['componentname' => $componentname]);
+        $DB->delete_records('tiny_elements_comp_variant', ['componentname' => $componentname]);
+        $DB->delete_records('tiny_elements_component', ['id' => $id]);
+
+        // Purge CSS and JS cache.
+        \tiny_elements\local\utils::purge_css_cache();
+        \tiny_elements\local\utils::rebuild_css_cache();
+        \tiny_elements\local\utils::purge_js_cache();
+        \tiny_elements\local\utils::rebuild_js_cache();
+    }
+
+    /**
+     * Duplicate a category.
+     *
+     * @param int $id
+     */
+    public function duplicate_compcat(int $id): void {
+        global $DB;
+        $compcat = $DB->get_record('tiny_elements_compcat', ['id' => $id]);
+        if ($compcat) {
+            $newcompcat = clone $compcat;
+            unset($newcompcat->id);
+            $newcompcat->displayname = get_string('copyof', 'tiny_elements', $newcompcat->displayname);
+            $newcompcat->name .= time();
```

Comment:
Is this intended? Adding that ugly timestamp? :) Probably yes, because it's not userfaced, right? Same goes for the other things (variants, components etc.)